e2e test error: Log in goroutine after TestNoNewEnemy has completed: no such process

When merging in the latest main changes (as of Jan 31  8:45am PT). I encountered the following error in the e2e test suite: 

```
2022-01-31T16:45:42.2426549Z     newenemy_test.go:343: 1643647542070212767 1643647542070212767.0000000039 -0.0000000039 1 1 3 3
2022-01-31T16:45:42.5098528Z     newenemy_test.go:343: 1643647542347004451 1643647542347004451.0000000037 -0.0000000037 1 1 3 3
2022-01-31T16:45:42.6102526Z === CONT  TestNoNewEnemy/protected_from_data_newenemy
2022-01-31T16:45:42.6102998Z     newenemy_test.go:231: spicedb is protected after 2 attempts
2022-01-31T16:45:42.6103560Z --- PASS: TestNoNewEnemy (80.27s)
2022-01-31T16:45:42.6103980Z     --- PASS: TestNoNewEnemy/protected_from_schema_newenemy (30.14s)
2022-01-31T16:45:42.6104444Z     --- PASS: TestNoNewEnemy/protected_from_data_newenemy (4.82s)
2022-01-31T16:45:42.6104687Z PASS
2022-01-31T16:45:42.6157997Z panic: Log in goroutine after TestNoNewEnemy has completed: no such process
2022-01-31T16:45:42.6158742Z 
2022-01-31T16:45:42.6158755Z 
2022-01-31T16:45:42.6158767Z 
2022-01-31T16:45:42.6159085Z goroutine 158 [running]:
2022-01-31T16:45:42.6159392Z testing.(*common).logDepth(0xc000106680, {0xc0002f0a98, 0x11}, 0x3)
2022-01-31T16:45:42.6159754Z 	/opt/hostedtoolcache/go/1.17.6/x64/src/testing/testing.go:779 +0x4c9
2022-01-31T16:45:42.6160020Z testing.(*common).log(...)
2022-01-31T16:45:42.6160316Z 	/opt/hostedtoolcache/go/1.17.6/x64/src/testing/testing.go:761
2022-01-31T16:45:42.6160620Z testing.(*common).Log(0x40ca07, {0xc0003fb8a0, 0xaf1be0, 0x1})
2022-01-31T16:45:42.6160936Z 	/opt/hostedtoolcache/go/1.17.6/x64/src/testing/testing.go:800 +0x45
2022-01-31T16:45:42.6161283Z github.com/authzed/spicedb/e2e.(*TLogger).Write(0xc00000e1b0, {0xc0001a6280, 0x10, 0x9})
2022-01-31T16:45:42.6161624Z 	/home/runner/work/spicedb/spicedb/e2e/util.go:17 +0xa6
2022-01-31T16:45:42.6161920Z fmt.Fprintln({0xd1d7a0, 0xc00000e1b0}, {0xc0003f37c0, 0x1, 0x1})
2022-01-31T16:45:42.6162248Z 	/opt/hostedtoolcache/go/1.17.6/x64/src/fmt/print.go:265 +0x75
2022-01-31T16:45:42.6162557Z github.com/authzed/spicedb/e2e.cleanupOnDone.func1()
2022-01-31T16:45:42.6162881Z 	/home/runner/work/spicedb/spicedb/e2e/run.go:55 +0x87
2022-01-31T16:45:42.6163164Z created by github.com/authzed/spicedb/e2e.cleanupOnDone
2022-01-31T16:45:42.6163476Z 	/home/runner/work/spicedb/spicedb/e2e/run.go:51 +0xa7
2022-01-31T16:45:42.6163788Z FAIL	github.com/authzed/spicedb/e2e/newenemy	80.288s
2022-01-31T16:45:42.6229755Z FAIL
2022-01-31T16:45:42.6269086Z ##[error]Process completed with exit code 1.
2022-01-31T16:45:42.6309006Z ##[group]Run actions/upload-artifact@v2
2022-01-31T16:45:42.6309229Z with:
2022-01-31T16:45:42.6309412Z   name: node-logs
2022-01-31T16:45:42.6309613Z   path: e2e/newenemy/*.log
2022-01-31T16:45:42.6309813Z   if-no-files-found: warn
``` 

The good news is this was a fast failure! Most of the previous failures were timeouts from a stalled test run.

--- a/e2e/newenemy/newenemy_test.go
+++ b/e2e/newenemy/newenemy_test.go
@@ -128,6 +128,9 @@ func TestNoNewEnemy(t *testing.T) {
 	require.NoError(t, vulnerableSpiceDb.Start(ctx, tlog, "vulnerable",
 		"--datastore-tx-overlap-strategy=insecure"))
 	require.NoError(t, vulnerableSpiceDb.Connect(ctx, tlog))
+	t.Cleanup(func() {
+		require.NoError(t, vulnerableSpiceDb.Stop(tlog))
+	})
 
 	t.Log("start protected spicedb cluster")
 	protectedSpiceDb := spice.NewClusterFromCockroachCluster(crdb,
@@ -139,6 +142,9 @@ func TestNoNewEnemy(t *testing.T) {
 		spice.WithDbName(dbName))
 	require.NoError(t, protectedSpiceDb.Start(ctx, tlog, "protected"))
 	require.NoError(t, protectedSpiceDb.Connect(ctx, tlog))
+	t.Cleanup(func() {
+		require.NoError(t, protectedSpiceDb.Stop(tlog))
+	})
 
 	t.Log("configure small ranges, single replicas, short ttl")
 	require.NoError(t, crdb.SQL(ctx, tlog, tlog,
@@ -150,7 +156,7 @@ func TestNoNewEnemy(t *testing.T) {
 	// seems to make the test converge faster
 	schemaData := generateSchemaData(4000, 500)
 	fillSchema(t, schemaTpl, schemaData, vulnerableSpiceDb[1].Client().V1Alpha1().Schema())
-	slowNodeId, err := crdb[1].NodeID(testCtx)
+	slowNodeId, err := crdb[1].NodeID(ctx)
 	require.NoError(t, err)
 
 	t.Log("modifying time")
@@ -302,13 +308,13 @@ func iterationsForHighConfidence(samples []int) (iterations int) {
 func checkDataNoNewEnemy(ctx context.Context, t testing.TB, schemaData []SchemaData, slowNodeId int, crdb cockroach.Cluster, spicedb spice.Cluster, maxAttempts int) (bool, int) {
 	prefix := prefixForNode(ctx, crdb[1].Conn(), schemaData, slowNodeId)
 	t.Log("filling with data to span multiple ranges for prefix", prefix)
-	fill(t, spicedb[0].Client().V0().ACL(), prefix, 500, 100)
+	fill(ctx, t, spicedb[0].Client().V0().ACL(), prefix, 500, 100)
 
 	for attempts := 1; attempts <= maxAttempts; attempts++ {
 		direct, exclude := generateTuple(prefix, objIdGenerator)
 
 		// write to node 1
-		r1, err := spicedb[0].Client().V0().ACL().Write(testCtx, &v0.WriteRequest{
+		r1, err := spicedb[0].Client().V0().ACL().Write(ctx, &v0.WriteRequest{
 			Updates: []*v0.RelationTupleUpdate{exclude},
 		})
 		require.NoError(t, err)
@@ -319,7 +325,7 @@ func checkDataNoNewEnemy(ctx context.Context, t testing.TB, schemaData []SchemaD
 		time.Sleep(100 * time.Millisecond)
 
 		// write to node 2 (clock is behind)
-		r2, err := spicedb[1].Client().V0().ACL().Write(testCtx, &v0.WriteRequest{
+		r2, err := spicedb[1].Client().V0().ACL().Write(ctx, &v0.WriteRequest{
 			Updates: []*v0.RelationTupleUpdate{direct},
 		})
 		require.NoError(t, err)
@@ -335,9 +341,9 @@ func checkDataNoNewEnemy(ctx context.Context, t testing.TB, schemaData []SchemaD
 		})
 		require.NoError(t, err)
 
-		r1leader, r2leader := getLeaderNode(testCtx, crdb[1].Conn(), exclude.Tuple), getLeaderNode(testCtx, crdb[1].Conn(), direct.Tuple)
-		ns1Leader := getLeaderNodeForNamespace(testCtx, crdb[1].Conn(), exclude.Tuple.ObjectAndRelation.Namespace)
-		ns2Leader := getLeaderNodeForNamespace(testCtx, crdb[1].Conn(), exclude.Tuple.User.GetUserset().Namespace)
+		r1leader, r2leader := getLeaderNode(ctx, crdb[1].Conn(), exclude.Tuple), getLeaderNode(ctx, crdb[1].Conn(), direct.Tuple)
+		ns1Leader := getLeaderNodeForNamespace(ctx, crdb[1].Conn(), exclude.Tuple.ObjectAndRelation.Namespace)
+		ns2Leader := getLeaderNodeForNamespace(ctx, crdb[1].Conn(), exclude.Tuple.User.GetUserset().Namespace)
 		z1, _ := zookie.DecodeRevision(r1.GetRevision())
 		z2, _ := zookie.DecodeRevision(r2.GetRevision())
 		t.Log(z1, z2, z1.Sub(z2).String(), r1leader, r2leader, ns1Leader, ns2Leader)
@@ -417,13 +423,13 @@ func checkSchemaNoNewEnemy(ctx context.Context, t testing.TB, schemaData []Schem
 		})))
 
 		// write the "direct" tuple
-		require.NoError(t, getErr(spicedb[0].Client().V0().ACL().Write(testCtx, &v0.WriteRequest{
+		require.NoError(t, getErr(spicedb[0].Client().V0().ACL().Write(ctx, &v0.WriteRequest{
 			Updates: []*v0.RelationTupleUpdate{direct},
 		})))
 
 		// write the "excludes" tuple
 		// writing to 1 primes the namespace cache on node 1 with the "allow" namespace
-		r2, err := spicedb[1].Client().V0().ACL().Write(testCtx, &v0.WriteRequest{
+		r2, err := spicedb[1].Client().V0().ACL().Write(ctx, &v0.WriteRequest{
 			Updates: []*v0.RelationTupleUpdate{exclude},
 		})
 		require.NoError(t, err)
@@ -431,7 +437,7 @@ func checkSchemaNoNewEnemy(ctx context.Context, t testing.TB, schemaData []Schem
 		// write the "exclude" schema. If this write hits the slow crdb node, it
 		// can get a revision in between the direct and exclude tuple writes
 		// which will cause check to fail, when it should succeed
-		require.NoError(t, getErr(spicedb[1].Client().V1Alpha1().Schema().WriteSchema(testCtx, &v1alpha1.WriteSchemaRequest{
+		require.NoError(t, getErr(spicedb[1].Client().V1Alpha1().Schema().WriteSchema(ctx, &v1alpha1.WriteSchemaRequest{
 			Schema: excludeSchema,
 		})))
 
@@ -442,7 +448,7 @@ func checkSchemaNoNewEnemy(ctx context.Context, t testing.TB, schemaData []Schem
 		checkAccess := func() bool {
 			var err error
 
-			canHas, err = spicedb[0].Client().V1().Permissions().CheckPermission(testCtx, &v1.CheckPermissionRequest{
+			canHas, err = spicedb[0].Client().V1().Permissions().CheckPermission(ctx, &v1.CheckPermissionRequest{
 				Consistency: &v1.Consistency{
 					Requirement: &v1.Consistency_AtExactSnapshot{AtExactSnapshot: zedtoken.NewFromRevision(rev)},
 				},
@@ -493,13 +499,13 @@ func BenchmarkBatchWrites(b *testing.B) {
 
 	directs, excludes := generateTuples("", b.N*20, objIdGenerator)
 	b.ResetTimer()
-	_, err := spicedb[0].Client().V0().ACL().Write(testCtx, &v0.WriteRequest{
+	_, err := spicedb[0].Client().V0().ACL().Write(ctx, &v0.WriteRequest{
 		Updates: excludes,
 	})
 	if err != nil {
 		b.Log(err)
 	}
-	_, err = spicedb[0].Client().V0().ACL().Write(testCtx, &v0.WriteRequest{
+	_, err = spicedb[0].Client().V0().ACL().Write(ctx, &v0.WriteRequest{
 		Updates: directs,
 	})
 	if err != nil {
@@ -517,24 +523,24 @@ func BenchmarkConflictingTupleWrites(b *testing.B) {
 	require.NoError(b, spicedb.Connect(ctx, tlog))
 
 	// fill with tuples to ensure we span multiple ranges
-	fill(b, spicedb[0].Client().V0().ACL(), "", 2000, 100)
+	fill(ctx, b, spicedb[0].Client().V0().ACL(), "", 2000, 100)
 
 	b.ResetTimer()
 
 	checkDataNoNewEnemy(ctx, b, generateSchemaData(1, 1), 1, crdb, spicedb, b.N)
 }
 
-func fill(t testing.TB, client v0.ACLServiceClient, prefix string, fillerCount, batchSize int) {
+func fill(ctx context.Context, t testing.TB, client v0.ACLServiceClient, prefix string, fillerCount, batchSize int) {
 	t.Log("filling prefix", prefix)
 	require := require.New(t)
 	directs, excludes := generateTuples(prefix, fillerCount, objIdGenerator)
 	for i := 0; i < fillerCount/batchSize; i++ {
 		t.Log("filling", i*batchSize, "to", (i+1)*batchSize)
-		_, err := client.Write(testCtx, &v0.WriteRequest{
+		_, err := client.Write(ctx, &v0.WriteRequest{
 			Updates: excludes[i*batchSize : (i+1)*batchSize],
 		})
 		require.NoError(err)
-		_, err = client.Write(testCtx, &v0.WriteRequest{
+		_, err = client.Write(ctx, &v0.WriteRequest{
 			Updates: directs[i*batchSize : (i+1)*batchSize],
 		})
 		require.NoError(err)

--- a/e2e/run.go
+++ b/e2e/run.go
@@ -4,6 +4,7 @@ import (
 	"context"
 	"fmt"
 	"io"
+	"os"
 	"os/exec"
 	"syscall"
 )
@@ -14,7 +15,7 @@ func Run(ctx context.Context, out, errOut io.Writer, args ...string) error {
 	if err != nil {
 		return err
 	}
-	cleanupOnDone(ctx.Done(), cmd.Process.Pid, errOut)
+	cleanupOnDone(ctx.Done(), cmd.Process.Pid, os.Stderr)
 	return cmd.Wait()
 }
 
@@ -25,7 +26,7 @@ func GoRun(ctx context.Context, out, errOut io.Writer, args ...string) (int, err
 		return 0, err
 	}
 	go func() {
-		cleanupOnDone(ctx.Done(), cmd.Process.Pid, errOut)
+		cleanupOnDone(ctx.Done(), cmd.Process.Pid, os.Stderr)
 
 		if err := cmd.Wait(); err != nil {
 			fmt.Fprintln(errOut, err)

--- a/e2e/run.go
+++ b/e2e/run.go
@@ -15,7 +15,7 @@ func Run(ctx context.Context, out, errOut io.Writer, args ...string) error {
 	if err != nil {
 		return err
 	}
-	cleanupOnDone(ctx.Done(), cmd.Process.Pid, os.Stderr)
+	cleanupOnDone(ctx.Done(), cmd.Process.Pid)
 	return cmd.Wait()
 }
 
@@ -26,7 +26,7 @@ func GoRun(ctx context.Context, out, errOut io.Writer, args ...string) (int, err
 		return 0, err
 	}
 	go func() {
-		cleanupOnDone(ctx.Done(), cmd.Process.Pid, os.Stderr)
+		cleanupOnDone(ctx.Done(), cmd.Process.Pid)
 
 		if err := cmd.Wait(); err != nil {
 			fmt.Fprintln(errOut, err)
@@ -48,12 +48,12 @@ func start(ctx context.Context, out, errOut io.Writer, args ...string) (*exec.Cm
 	return cmd, nil
 }
 
-func cleanupOnDone(done <-chan struct{}, pid int, errOut io.Writer) {
+func cleanupOnDone(done <-chan struct{}, pid int) {
 	go func() {
 		<-done
 		// negative Pid sends the signal to all child processes in the group
 		if err := syscall.Kill(-pid, syscall.SIGKILL); err != nil {
-			fmt.Fprintln(errOut, err)
+			fmt.Fprintln(os.Stderr, err)
 		}
 	}()
 }

